### PR TITLE
bpo-34880: assert statement misbehavior if AssertionError is shadowed

### PR DESCRIPTION
**Pull request by ZackerySpytz**
The assert statement was not working properly if the AssertionError
exception was being shadowed.

https://bugs.python.org/issue34880

### CODE REVIEW

File: Lib/importlib/_bootstrap_external.py
Comment (serhiy-storchaka) [inline Lib/importlib/_bootstrap_external.py]: I think it is worth to increase the magic number more with a new version. To 3420.

File: Lib/test/test_exceptions.py
Comment (serhiy-storchaka) [inline Lib/test/test_exceptions.py]: Why use `globals()['AssertionError']` instead of just setting `AssertionError`?

File: Lib/test/test_exceptions.py
Comment (serhiy-storchaka) [inline Lib/test/test_exceptions.py:1297]: You need to remove it also if no exception is raised, or non-Exception exception is raised.

File: Lib/test/test_exceptions.py
Comment (serhiy-storchaka) [inline Lib/test/test_exceptions.py:1295]: This test will fail when run in optimized mode (`-O`).

File: Python/ceval.c
Comment (serhiy-storchaka) [inline Python/ceval.c]: Why convert the argument to `str`?

**Comment by markshannon** [comment]
I'd like to keep any new bytecodes as simple as possible. The simplest I can think of is a `LOAD_ASSERTION_ERROR` which would be equivalent to `LOAD_CONSTANT AssertionError`.
It would also keep the changes to the compiler to a minimum.

**Comment by serhiy-storchaka** [comment]
I concur with @markshannon. There is no benefit from implementing it as a single bytecode: it is neither performance critical nor should be atomical. Using `LOAD_ASSERTION_ERROR` would simplify the compiler and the bytecode interpreter a bit.

File: Lib/importlib/_bootstrap_external.py
Comment (ZackerySpytz) [reply Lib/importlib/_bootstrap_external.py]: Okay. Done.

File: Lib/test/test_exceptions.py
Comment (ZackerySpytz) [reply Lib/test/test_exceptions.py:1297]: You are right.

File: Lib/test/test_exceptions.py
Comment (ZackerySpytz) [reply Lib/test/test_exceptions.py:1295]: It does not fail, but it silently passes without testing anything. `test_assert_with_tuple_arg()` in test_raise.py tests `assert` in the same way. I added a `skipUnless` decorator.

File: Lib/test/test_exceptions.py
Comment (ZackerySpytz) [reply Lib/test/test_exceptions.py]: The bug involves `AssertionError` being shadowed at the global level, no?

**Comment by ZackerySpytz** [comment]
Thanks for all of the comments. I have implemented `LOAD_ASSERTION_ERROR`. Please take a look.

File: Python/compile.c
Comment (ZackerySpytz) [inline Python/compile.c]: Ah, this should be 1, not 0.

File: Lib/test/test_exceptions.py
Comment (serhiy-storchaka) [reply Lib/test/test_exceptions.py]: I meant that you can write
```python
global AssertionError
AssertionError = TypeError
```

File: Lib/test/test_exceptions.py
Comment (serhiy-storchaka) [reply Lib/test/test_exceptions.py:1295]: You can add also a test for
```python
exec(compile('assert False, "hello"', 'test', 'exec', optimize=0), {'AssertionError': TypeError})
```
which can work even in optimized mode.

**Review by fusioncid — COMMENTED** [review]
It's better for "{" at the end of the previous line.

File: Lib/test/test_exceptions.py
Comment (ZackerySpytz) [reply Lib/test/test_exceptions.py]: Okay. I will make that change.

**Comment by ZackerySpytz** [comment]
@qigangxu All of the code in this PR follows the [PEP 7](https://www.python.org/dev/peps/pep-0007/) and [PEP 8](https://www.python.org/dev/peps/pep-0008/) style guides, so I'm not sure what the purpose of your comment is.

File: Lib/test/test_exceptions.py
Comment (ZackerySpytz) [reply Lib/test/test_exceptions.py:1295]: @serhiy-storchaka Should that line go in its own test method?

Comment (serhiy-storchaka) [reply Lib/test/test_exceptions.py:1295]: If you prefer.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM, but it needs to wait until #15320 be merged.

**Comment by qigangxu** [comment]
> @qigangxu All of the code in this PR follows the [PEP 7](https://www.python.org/dev/peps/pep-0007/) and [PEP 8](https://www.python.org/dev/peps/pep-0008/) style guides, so I'm not sure what the purpose of your comment is.

Sorry, no code location information was provided last comment. The issue is in line 3327 of Python/compile.c.

**Review by serhiy-storchaka — APPROVED** [review]

**Pull request merged**
